### PR TITLE
mempool: p2p refactor

### PR DESCRIPTION
**Pull request by alexanderbez**
## Description

Prepares the mempool reactor for the newly designed p2p changes per [ADR 062](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-062-p2p-architecture.md).

ref: #5670

### CODE REVIEW

**Review by erikgrinaker — APPROVED** [review]
Looks great!

File: libs/sync/deadlock.go [hunk 1]
```diff
@@ -15,3 +15,31 @@ type Mutex struct {
 type RWMutex struct {
 	deadlock.RWMutex
 }
+
+// Closer implements a primitive to close a channel that signals process
+// termination while allowing a caller to call Close multiple times safely. It
+// should be used in cases where guarantees cannot be made about when and how
+// many times closure is executed.
+type Closer struct {
+	closeOnce deadlock.Once
```
[the remark this answers is not in the document]
Comment (alexanderbez) [reply libs/sync/deadlock.go:24]: Not sure I follow?

[the remark this answers is not in the document]
Comment (erikgrinaker) [reply libs/sync/deadlock.go:24]: I was thinking we could use an interface or something, and just do e.g. `var Once OnceInterface = deadlock.Once`, and same in a different file with `sync.Once`. However, turns out `deadlock.Once` doesn't do anything so we can just move it to a separate file and don't use `deadlock.Once` at all.

https://github.com/sasha-s/go-deadlock/blob/1595213edefa28ca5047b00340c63557f9c051d0/deadlock.go#L58

**Comment by erikgrinaker** [comment]
Looks like the E2E tests are failing here btw.

**Comment by alexanderbez** [comment]
> Looks like the E2E tests are failing here btw.

Yes, because I missed adding an explicit `height > 0` check (hence the timeout because `height` will always be zero until we have everything setup). Fixed it and now it all passes.

**Pull request merged**